File: archive/paperclip/agents/audit/chain.py
"""
Per-(tenant, run) hash chain (ADR-0001 §D2).

`record_hash = SHA256(canonical_json(event w/o recordHash) || prev_hash)`

The chain head is the (tenant_id, run_id) pair.  The verifier walks
one run at a time; a run with N events costs O(N) to verify.  A
deleted or redacted event surfaces as a chain break, not a silent
verification failure: the store emits a synthetic `event_redacted`
event so the head always advances.  See `test_chain.py`.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass
from typing import Iterable, List, Tuple

from .schema import AuditEvent


# 32 zero bytes, hex-encoded.  The `prev_hash` of the first event in
# any (tenant, run) chain.  Using a sentinel of fixed length keeps
# the canonical hash input stable.
GENESIS_HASH = "0" * 64


@dataclass
class ChainBreak:
    """One place where the chain does not match.  The verifier
    returns a list of these rather than raising so the board can
    see every break, not just the first."""
    index: int
    event_id: str
    expected_prev_hash: str
    actual_prev_hash: str
    reason: str         # "prev_hash_mismatch" | "self_hash_mismatch" | "missing_event"

    def to_dict(self) -> dict:
        return {
            "index": self.index,
            "eventId": self.event_id,
            "expectedPrevHash": self.expected_prev_hash,
            "actualPrevHash": self.actual_prev_hash,
            "reason": self.reason,
        }


class HashChain:
    """Stateless chain walker.  Use `verify(events)` to walk a
    run; use `next_hash(event, prev_hash)` to compute the head."""

    @staticmethod
    def next_hash(event: AuditEvent, prev_hash: str) -> str:
        """Compute the record hash for `event` chained to `prev_hash`."""
        payload = event.canonical_bytes() + b"|" + prev_hash.encode("utf-8")
        return hashlib.sha256(payload).hexdigest()

    @staticmethod
    def verify(events: Iterable[AuditEvent]) -> Tuple[bool, List[ChainBreak]]:
        """Walk `events` in order and return (ok, breaks).  `ok` is
        True iff no breaks were found.  The caller is responsible
        for delivering `events` in append order for the (tenant,
        run) pair; the verifier does not sort."""
        breaks: List[ChainBreak] = []
        prev = GENESIS_HASH
        for idx, ev in enumerate(events):
            if ev.prev_hash != prev:
                breaks.append(ChainBreak(
                    index=idx,
                    event_id=ev.event_id,
                    expected_prev_hash=prev,
                    actual_prev_hash=ev.prev_hash,
                    reason="prev_hash_mismatch",
                ))
                # Don't trust any subsequent hashes from this point.
                # We continue walking so the board can see the full
                # picture, but we don't update `prev` -- the chain is
                # broken here.
                continue
            expected = HashChain.next_hash(ev, ev.prev_hash)
            if ev.record_hash != expected:
                breaks.append(ChainBreak(
                    index=idx,
                    event_id=ev.event_id,
                    expected_prev_hash=prev,
                    actual_prev_hash=ev.prev_hash,
                    reason="self_hash_mismatch",
                ))
                continue
            prev = ev.record_hash
        return (len(breaks) == 0), breaks

File: packages/connector-events/src/chain.ts
/**
 * Per-(tenant, binding) hash chain — FORA-484 AC #5.
 *
 * The connector event chain is distinct from the audit-spine chain
 * (`agents/audit/chain.py`), which is per-(tenant, run). Connector
 * chains are per-(tenant, binding) — every credential binding gets
 * its own append-only chain, so a binding rotation cleanly terminates
 * one chain and starts another.
 *
 *   event_hash = SHA-256(canonical_json(event w/o audit_chain.event_hash)
 *                         || ":" || prev_event_hash)
 *
 * The chain head is the (tenant_id, binding_id) pair. A break is
 * reported as a `ChainBreak` record so the verifier can return every
 * problem in one pass, not just the first.
 */

import { createHash } from 'node:crypto';
import type { AuditChain, ConnectorEvent } from './envelope.js';

/** 32 zero bytes, hex-encoded. The `prev_event_hash` of the first event. */
export const GENESIS_HASH = '0'.repeat(64);

/** Sentinel returned by `nextHash` when the canonical form cannot be computed. */
export class ChainIntegrityError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ChainIntegrityError';
  }
}

/**
 * Canonical JSON for hashing. Recursively sorts object keys so the
 * output is stable regardless of insertion order. `undefined` is
 * omitted (mirrors the Python `canonical_json` precedent). NaN/Inf
 * throw via `JSON.stringify`'s default behaviour.
 */
export function canonicalJson(obj: unknown): string {
  return JSON.stringify(sortKeysDeep(obj));
}

function sortKeysDeep(value: unknown): unknown {
  if (value === null || typeof value !== 'object') return value;
  if (Array.isArray(value)) return value.map(sortKeysDeep);
  const entries = Object.entries(value as Record<string, unknown>)
    .filter(([, v]) => v !== undefined)
    .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0));
  const out: Record<string, unknown> = {};
  for (const [k, v] of entries) out[k] = sortKeysDeep(v);
  return out;
}

/** Hex SHA-256 of any JSON-serialisable payload. Stable across processes. */
export function digestOf(payload: unknown): string {
  return createHash('sha256').update(canonicalJson(payload), 'utf8').digest('hex');
}

/**
 * Compute the record hash for `event` chained to `prevHash`.
 *
 * The input is `canonical_json(stableEvent) || ":" || prevHash`.
 * `stableEvent` is the event with `audit_chain.event_hash` removed
 * (the field being computed).
 */
export function nextHash(event: ConnectorEvent, prevHash: string): string {
  if (!/^[0-9a-f]{64}$/.test(prevHash)) {
    throw new ChainIntegrityError(`prevHash must be a 64-char hex digest: ${prevHash}`);
  }
  const stable = stripEventHash(event);
  const input = `${canonicalJson(stable)}:${prevHash}`;
  return createHash('sha256').update(input, 'utf8').digest('hex');
}

/** Return the event with `audit_chain.event_hash` removed (kept immutable). */
export function stripEventHash(event: ConnectorEvent): Omit<ConnectorEvent, 'audit_chain'> & {
  audit_chain: Omit<AuditChain, 'event_hash'>;
} {
  const { audit_chain, ...rest } = event;
  const { event_hash: _eh, ...chain } = audit_chain;
  return { ...rest, audit_chain: chain };
}

export interface ChainBreak {
  index: number;
  event_id: string;
  expected_prev_hash: string;
  actual_prev_hash: string;
  reason: 'prev_hash_mismatch' | 'self_hash_mismatch';
}

/** Walk `events` and verify the chain. Returns (ok, breaks). */
export function verifyChain(events: ConnectorEvent[]): {
  ok: boolean;
  breaks: ChainBreak[];
} {
  const breaks: ChainBreak[] = [];
  let prev = GENESIS_HASH;
  for (let idx = 0; idx < events.length; idx += 1) {
    const ev = events[idx]!;
    if (ev.audit_chain.prev_event_hash !== prev) {
      breaks.push({
        index: idx,
        event_id: ev.event_id,
        expected_prev_hash: prev,
        actual_prev_hash: ev.audit_chain.prev_event_hash,
        reason: 'prev_hash_mismatch',
      });
      // Don't trust any subsequent hashes from this point.
      continue;
    }
    const expected = nextHash(ev, ev.audit_chain.prev_event_hash);
    if (ev.audit_chain.event_hash !== expected) {
      breaks.push({
        index: idx,
        event_id: ev.event_id,
        expected_prev_hash: prev,
        actual_prev_hash: ev.audit_chain.prev_event_hash,
        reason: 'self_hash_mismatch',
      });
      continue;
    }
    prev = ev.audit_chain.event_hash;
  }
  return { ok: breaks.length === 0, breaks };
}

/** Generate a stable event id: `evt-<uuid16>`. */
export function makeEventId(): string {
  // crypto.randomUUID gives 32 hex chars; we keep the 16-char convention.
  const hex = createHash('sha256')
    .update(`${Date.now()}:${Math.random()}:${process.pid}`)
    .digest('hex')
    .slice(0, 32);
  return `evt-${hex}`;
}